src/algebra/symmetrized.lean

```diff
@@ -0,0 +1,211 @@
+unsym_injective.add_comm_semigroup _ (λ _ _, rfl)
+
+instance [add_monoid α] : add_monoid (αˢʸᵐ) :=
+unsym_injective.add_monoid_smul _ rfl (λ _ _, rfl) (λ _ _, rfl)
+
+instance [add_group α] : add_group (αˢʸᵐ) :=
+unsym_injective.add_group_smul _ rfl
+  (λ _ _, rfl) (λ _, rfl) (λ _ _, rfl) (λ _ _, rfl) (λ _ _, rfl)
+
+instance [add_comm_monoid α] : add_comm_monoid (αˢʸᵐ) :=
+{ ..sym_alg.add_comm_semigroup, ..sym_alg.add_monoid }
+
+instance [add_comm_group α] : add_comm_group (αˢʸᵐ) :=
+{ ..sym_alg.add_comm_monoid, ..sym_alg.add_group }
+
+instance {R : Type*} [semiring R] [add_comm_monoid α] [module R α] : module R αˢʸᵐ :=
+function.injective.module R ⟨unsym, rfl, λ _ _, rfl⟩ (λ _ _, id) (λ _ _, rfl)
+
+lemma mul_def [ring α] [invertible (2 : α)] (a b : αˢʸᵐ) :
+  a * b = sym (⅟2*(unsym a * unsym b + unsym b * unsym a)) := by refl
+
+instance [has_mul α] [has_add α] [has_one α] [invertible (2 : α)] (a : α) [invertible a] :
+  invertible (sym a) :=
+{ inv_of := sym ⅟a,
+  inv_of_mul_self := begin
+    change ⅟2 * (⅟a * a + a * ⅟a) = 1,
+    rw [mul_inv_of_self, inv_of_mul_self],
+    exact mul_inv_of_self (⅟ 2),
+  end,
+  mul_inv_of_self := begin
+    change ⅟2 * (a* ⅟a + ⅟a * a)=1,
+    rw [mul_inv_of_self, inv_of_mul_self],
+    exact mul_inv_of_self (⅟ 2),
+  end }
+
+lemma unsym_mul [has_mul α] [has_add α] [has_one α] [invertible (2 : α)] (a b : αˢʸᵐ) :
+  unsym (a * b) = ⅟2*(unsym a * unsym b + unsym b * unsym a) := by refl
+
+lemma sym_mul_sym [has_mul α] [has_add α] [has_one α] [invertible (2 : α)] (a b : α) :
+  sym a * sym b = sym (⅟2*(a * b + b * a)) :=
+rfl
+
+/- The symmetrization of a real (unital, associative) algebra is a non-associative ring -/
+instance [ring α] [invertible (2 : α)] : non_unital_non_assoc_ring (αˢʸᵐ) :=
+{ zero_mul := λ _,
+  begin
+    rw [mul_def, unsym_zero, zero_mul, mul_zero, add_zero, mul_zero, sym_zero],
+    exact rfl,
+  end,
+  mul_zero :=  λ _,
+  begin
+    rw [mul_def, unsym_zero, zero_mul, mul_zero, add_zero, mul_zero, sym_zero],
+    exact rfl,
+  end,
+  left_distrib := λ a b c, match a, b, c with
+    | sym a, sym b, sym c := begin
+      rw [sym_mul_sym, sym_mul_sym, ←sym_add, sym_mul_sym, ←sym_add, mul_add a, add_mul _ _ a,
+        add_add_add_comm, mul_add],
+    end
+  end,
+  right_distrib := λ a b c, match a, b, c with
+    | sym a, sym b, sym c := begin
+      rw [sym_mul_sym, sym_mul_sym, ←sym_add, sym_mul_sym, ←sym_add, mul_add c, add_mul _ _ c,
+        add_add_add_comm, mul_add],
+    end
+  end,
+  ..sym_alg.has_mul,
+  ..sym_alg.add_comm_group, }
+
+/- The squaring operation coincides for both multiplications -/
+lemma unsym_mul_self [ring α] [invertible (2 : α)] (a : αˢʸᵐ) : unsym (a*a) = unsym a * unsym a :=
+begin
+  rw [mul_def, unsym_sym],
+  abel,
+  simp only [int.cast_bit0, int.cast_one, inv_of_mul_self_assoc, zsmul_eq_mul],
+end
+
+lemma sym_mul_self [ring α] [invertible (2 : α)] (a : α) : sym a*a = sym a * sym a :=
```

Comment:
```suggestion
lemma sym_mul_self [ring α] [invertible (2 : α)] (a : α) : sym (a*a) = sym a * sym a :=
```
What you have at the moment is nonsense and multiplies a sym and nonsym element

Reply:
Yes, sorry. If I enter:
```
variables [ring α] (a : α) 

#check (sym a) * a
```
Then lean produces the error:
```
failed to synthesize type class instance for
α : Type u_1,
_inst_1 : ring α,
a : α
⊢ has_mul αˢʸᵐ
```
So I'm surprised that lean seems to have accepted a nonsense statement and proof without producing any error or warning. Is this something we should be concerned about?

Reply:
This would be fixed by doing `structure sym_alg` instead of `def sym_alg`, but that has its own technical drawbacks.